lib/elixir/lib/partition_supervisor.ex

```diff
@@ -0,0 +1,315 @@
+defmodule PartitionSupervisor do
+  @moduledoc """
+  A supervisor that starts multiple partitions of the same child.
+
+  Certain processes may become bottlenecks in large systems.
+  If those processes can have their state trivially partitioned,
+  in a way there is no dependency between them, then they can use
+  the `PartitionSupervisor` to create multiple isolated and
+  independent partitions.
+
+  Once the `PartitionSupervisor` starts, you can dispatch to its
+  children using `{:via, PartitionSupervisor, {name, key}}`, where
+  `name` is the name of the `PartitionSupervisor` and key is used
+  for routing.
+
+  ## Example
+
+  The `DynamicSupervisor` is a single process responsible for starting
+  other processes. In some applications, the `DynamicSupervisor` may
+  become a bottleneck. To address this, you can start multiple instances
+  of the `DynamicSupervisor` and then pick a "random" instance to start
+  the child on.
+
+  Instead of:
+
+      children = [
+        {DynamicSupervisor, name: MyApp.DynamicSupervisor}
+      ]
+
+      Supervisor.start_link(children, strategy: :one_for_one)
+
+  and:
+
+      DynamicSupervisor.start_child(MyApp.DynamicSupervisor, {Agent, fn -> %{} end})
+
+  You can do this:
+
+      children = [
+        {PartitionSupervisor,
+         child_spec: DynamicSupervisor,
+         name: MyApp.DynamicSupervisors}
+      ]
+
+      Supervisor.start_link(children, strategy: :one_for_one)
+
+  and then:
+
+      DynamicSupervisor.start_child(
+        {:via, PartitionSupervisor, {MyApp.DynamicSupervisors, self()}},
+        {Agent, fn -> %{} end}
+      )
+
+  In the code above, we start a partition supervisor that will by default
+  start a dynamic supervisor for each core in your machine. Then, instead
+  of calling the `DynamicSupervisor` by name, you call it through the
+  partition supervisor using the `{:via, PartitionSupervisor, {name, key}}`
+  format. We picked `self()` as the routing key, which means each process
+  will be assigned one of the existing dynamic supervisors. See `start_link/1`
+  to see all options supported by the `PartitionSupervisor`.
+
+  ## Implementation notes
+
+  The `PartitionSupervisor` requires a name as an atom to be given on start,
```

Comment:
This is a bit limiting, though it will probably suffice for most cases. I find it a bit disappointing that I can't start this from e.g. a child of a dynamic supervisor, or any other process which is started in unknown number of copies.

Reply:
We could make it generic by using a registry and storing the ETS table there. We would need to assemble the Registry dynamically though but that's not a problem. IIRC many OTP services get started on first use. it is doable and we can do it in the future. :)